[1/N] Replace data_ptr with mutable_data_ptr

cyyever opened the pull request:
This PR replaces `data_ptr` with `mutable_data_ptr` because they are aliases and the change enhances code readability.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

cyyever commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -2056,7 +2056,7 @@ void scaled_gemm(
       auto pointer_mode = CUBLASLT_POINTER_MODE_DEVICE;
       computeDesc.setAttribute(CUBLASLT_MATMUL_DESC_POINTER_MODE, pointer_mode);
-      alpha_ptr = user_alpha.data_ptr<float>();
+      alpha_ptr = user_alpha.mutable_data_ptr<float>();
```
See the comment above the method definition:
```

  // TODO(#97856) Make this return a const pointer. This currently
  //              returns a non-const pointer because of the large
  //              number of clients that we still want to audit before
  //              migrating to mutable_data_ptr().
  void* data_ptr() const {
    return mutable_data_ptr();
  }
```
`mutable_data_ptr` and `const_data_ptr` were introduced to replace `data_ptr` usage.

thenumberouscode replied:
Thanks

Skylion007 reviewed (commented):
So the reason we haven't done a code-mod automated change is we want to change all these to const_data ptr we can't even if the temp auto pointer is non-const and gets cast back.

We should double check what pointers can actually be const.

cyyever commented on aten/src/ATen/native/AdaptiveMaxPooling3d.cpp:
```diff
@@ -312,8 +312,8 @@ TORCH_IMPL_FUNC(adaptive_max_pool3d_out_cpu)
-          auto output_data = output.data_ptr<scalar_t>();
-          auto indices_data = indices.data_ptr<int64_t>();
+          auto output_data = output.mutable_data_ptr<scalar_t>();
+          auto indices_data = indices.mutable_data_ptr<int64_t>();
```
Fixed

cyyever commented on aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:
```diff
@@ -108,8 +108,8 @@ void apply_cholesky_inverse(Tensor& input, Tensor& infos, bool upper) {
 
-  auto input_data = input.data_ptr<scalar_t>();
-  auto infos_data = infos.data_ptr<int>();
+  auto input_data = input.mutable_data_ptr<scalar_t>();
```
Fixed

cyyever commented on aten/src/ATen/native/AdaptiveMaxPooling3d.cpp:
```diff
@@ -335,8 +335,8 @@ TORCH_IMPL_FUNC(adaptive_max_pool3d_out_cpu)
-          auto output_data = output.data_ptr<scalar_t>();
-          auto indices_data = indices.data_ptr<int64_t>();
+          auto output_data = output.mutable_data_ptr<scalar_t>();
+          auto indices_data = indices.mutable_data_ptr<int64_t>();
```
indices_data is indeed modified.

cyyever committed:
More fixes

Signed-off-by: Yuanyuan Chen <cyyever@outlook.com>

Skylion007 commented on aten/src/ATen/native/AdaptiveAveragePooling3d.cpp:
```diff
@@ -252,7 +252,7 @@ Tensor& adaptive_avg_pool3d_backward_out_cpu_template(
         input.scalar_type(), "adaptive_avg_pool3d_backward_cpu", [&] {
           /* get raw pointers */
-          scalar_t* gradInput_data = gradInput.data_ptr<scalar_t>();
+          scalar_t* gradInput_data = gradInput.mutable_data_ptr<scalar_t>();
```
Why is the input data mutable here?

albanD reviewed (commented):
Note that this has real impact, in particular in the COW Tensor case.
cc @amjames can someone from that workstream take a quick look here?

kurtamohler commented:
When changing a particular `data_ptr` to either `mutable_data_ptr` or `const_data_ptr`, the surrounding code should ideally be carefully examined for const-correctness with COW behavior in mind. The `data_ptr` name is a nice marker to help us find call sites that haven't been properly vetted yet, and IMO that benefit is worth the slightly decreased code readability.

I notice that this PR doesn't add any changes to tests. There are existing opinfos tests to check correct COW behavior, like avoiding mutations of all COW data and materializing when expected. I'm not sure if the existing tests exercise all of the changes in this PR, but I wouldn't be surprised if there are some call sites in the code base that aren't covered.

In the past, I've examined some parts of the code base to make data pointer usage const-correct, avoid unnecessary COW materializations, and avoid mutating COW data, and I saw a few different situations in which the `data_ptr` -> `const_data_ptr` or -> `mutable_data_ptr` change that seemed appropriate at surface level would have been wrong, and more changes to the surrounding code were actually required. I don't remember all of these situations, but to give two examples:

* A pointer which seems like it can be marked const may sometimes get nonconst-casted later on. In these cases, a `data_ptr` -> `const_data_ptr` change would avoid COW materialization, but then the cast later on makes it unsafe at best and at worst the COW data can actually get mutated.

* When a `data_ptr` -> `mutable_data_ptr` change seems required at surface level, sometimes it can actually be changed to `const_data_ptr` to avoid an unnecessary COW materialize, but only after fixing const-correctness of other related code.

If this PR accounts for all of these concerns, then I'm happy with that. But if not, and if this PR is just a systematic renaming without full understanding of the context of each change, then I'd suggest it would be better to keep the `data_ptr` calls until they are examined in more detail.

cyyever commented:
@kurtamohler Thank you for the detailed consideration. These changes are not simply grepping and renaming. Each change is checked with its code context. However, because our test coverage is not 100%, there can be unnoticed errors. 
Let's consider the two mentioned incorrect cases:
1. `data_ptr -> mutable_data_ptr` but can be actually `const`. In this case the system wouldn't break, but the optimisation chance is hidden. I plan to eliminate these possibilities in this PR by checking them again. Also note that sometimes a CUDA API (API design issue) requires mutable pointers but they are indeed const according to its semantics and from the documentation.
2. `data_ptr -> const_data_ptr` but is indeed used as mutable by casting in some future code path. Frankly speaking, We reply on the C++ typing system and the current tests to avoid these risks. That is why the tests are not touched in this PR. I will check each such change in the PR again.

These thoughts show my understanding to the problem. Nevertheless, this PR could be a good starting point towards classifying data_ptr usage.

pearu commented on aten/src/ATen/native/PackedSequence.cpp:
```diff
@@ -38,7 +38,7 @@ std::tuple<Tensor, Tensor> _pack_padded_sequence(const Tensor& _input, const Ten
 
   int64_t batch_size = input.size(1);
-  int64_t * lengths = lengths_t.data_ptr<int64_t>();
+  int64_t * lengths = lengths_t.mutable_data_ptr<int64_t>();
```
```suggestion
  const int64_t * lengths = lengths_t.const_data_ptr<int64_t>();
```

pearu commented on aten/src/ATen/native/RowwisePrune.cpp:
```diff
@@ -36,10 +36,10 @@ std::tuple<Tensor, Tensor> _rowwise_prune_helper(
-        compressed_indices_mapping.data_ptr<input_t>();
-    auto weights_data = weights.data_ptr<scalar_t>();
+        compressed_indices_mapping.mutable_data_ptr<input_t>();
+    const auto weights_data = weights.const_data_ptr<scalar_t>();
```
Nit, just for consistency:
```c++
    auto* compressed_indices_mapping_data =
        compressed_indices_mapping.mutable_data_ptr<input_t>();
    const auto* weights_data = weights.const_data_ptr<scalar_t>();
```

pearu commented on aten/src/ATen/native/miopen/RNN_miopen.cpp:
```diff
@@ -905,7 +905,7 @@ std::pair<Tensor, hidden_type> _miopen_impl(
 
     TORCH_CHECK(_batch_sizes.dim() == 1, "batch_sizes tensor should be 1D");
-    IntArrayRef batch_sizes { _batch_sizes.data_ptr<int64_t>(), static_cast<size_t>(_batch_sizes.size(0)) };
+    IntArrayRef batch_sizes { _batch_sizes.mutable_data_ptr<int64_t>(), static_cast<size_t>(_batch_sizes.size(0)) };
```
```suggestion
    IntArrayRef batch_sizes { _batch_sizes.const_data_ptr<int64_t>(), static_cast<size_t>(_batch_sizes.size(0)) };
```

pearu commented on aten/src/ATen/native/mkldnn/Linear.cpp:
```diff
@@ -438,12 +438,12 @@ Tensor mkl_linear(
-    auto weight_ptr = (float*)(w.get_data_handle());
-    auto out_ptr = output.data_ptr<float>();
+    auto in_ptr = self_.const_data_ptr<float>();
+    auto weight_ptr = (const float*)(w.get_data_handle());
```
```suggestion
    auto weight_ptr = reinterpret_cast<const float*>(w.get_data_handle());
```

pearu commented on aten/src/ATen/native/nested/NestedTensorBinaryOps.cpp:
```diff
@@ -58,8 +58,8 @@ static get_elementwise_nested_tensor_impl(
-  int64_t *self_offsets_ptr = self_offsets.data_ptr<int64_t>();
-  int64_t *other_offsets_ptr = other_ptr->get_storage_offsets().data_ptr<int64_t>();
+  int64_t *self_offsets_ptr = self_offsets.mutable_data_ptr<int64_t>();
+  int64_t *other_offsets_ptr = other_ptr->get_storage_offsets().mutable_data_ptr<int64_t>();
```
```suggestion
  const int64_t *self_offsets_ptr = self_offsets.const_data_ptr<int64_t>();
  const int64_t *other_offsets_ptr = other_ptr->get_storage_offsets().const_data_ptr<int64_t>();
```

pearu commented on aten/src/ATen/native/nested/NestedTensorUtils.cpp:
```diff
@@ -88,7 +88,7 @@ std::vector<Tensor> chunk_nested_tensor(const Tensor& self, int64_t chunks, int6
   const auto& strides = self_impl->get_nested_strides();
   const auto offsets = self_impl->get_storage_offsets();
-  int64_t *offsets_ptr = offsets.data_ptr<int64_t>();
+  int64_t *offsets_ptr = offsets.mutable_data_ptr<int64_t>();
```
```suggestion
  const int64_t *offsets_ptr = offsets.const_data_ptr<int64_t>();
```

pearu commented on aten/src/ATen/native/nested/NestedTensorUtils.cpp:
```diff
@@ -143,7 +143,7 @@ std::vector<Tensor> split_with_sizes_nested(
   const auto& strides = self_impl->get_nested_strides();
   const auto offsets = self_impl->get_storage_offsets();
-  int64_t *offsets_ptr = offsets.data_ptr<int64_t>();
+  int64_t *offsets_ptr = offsets.mutable_data_ptr<int64_t>();
```
```suggestion
  const int64_t *offsets_ptr = offsets.const_data_ptr<int64_t>();
```

pearu commented on aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp:
```diff
@@ -18,7 +18,7 @@ int64_t get_nnz(const Tensor& nestedtensor) {
   auto size_tensor_stride = sizes.stride(0);
   const int64_t batch_size = nestedtensor.size(0);
-  auto* sizes_ptr = sizes.data_ptr<int64_t>();
+  auto* sizes_ptr = sizes.mutable_data_ptr<int64_t>();
```
```suggestion
  const auto* sizes_ptr = sizes.const_data_ptr<int64_t>();
```

pearu commented on aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp:
```diff
@@ -49,8 +49,8 @@ int64_t get_nnz(const Tensor& nestedtensor) {
 
-    auto* sizes_ptr = sizes.data_ptr<int64_t>();
-    auto* cumulative_seqlen_ptr = cumulative_seqlen.data_ptr<int32_t>();
+    auto* sizes_ptr = sizes.mutable_data_ptr<int64_t>();
```
```suggestion
    const auto* sizes_ptr = sizes.const_data_ptr<int64_t>();
```

pearu commented on aten/src/ATen/native/sparse/SparseBlasImpl.cpp:
```diff
@@ -365,28 +365,28 @@ void addmv_out_sparse_csr(
-        mat.col_indices().data_ptr<idx_t>(),
+    addmv_sparse_csr(cont_values.mutable_data_ptr<scalar_t>(),
+        mat.crow_indices().mutable_data_ptr<idx_t>(),
+        mat.col_indices().mutable_data_ptr<idx_t>(),
```
```suggestion
    addmv_sparse_csr(cont_values.const_data_ptr<scalar_t>(),
        mat.crow_indices().const_data_ptr<idx_t>(),
        mat.col_indices().const_data_ptr<idx_t>(),
```

pearu commented on aten/src/ATen/native/sparse/SparseMatMul.cpp:
```diff
@@ -206,22 +206,22 @@ void sparse_matmul_kernel(
       mat2_csr.col_indices().stride(-1));
   auto mat2_values_ptr = StridedRandomAccessor<scalar_t>(
-      mat2_csr.values().data_ptr<scalar_t>(),
+      mat2_csr.values().mutable_data_ptr<scalar_t>(),
```
These all should use `const_data_ptr` with appropriate `const_cast`'s.

pearu commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -461,9 +461,9 @@ static SparseTensor& add_out_sparse_contiguous(SparseTensor& r, const SparseTens
-          scalar_t* s_values_ptr = s_values.data_ptr<scalar_t>();
-          scalar_t* r_values_ptr = r_values.data_ptr<scalar_t>();
+          scalar_t* t_values_ptr = t_values.mutable_data_ptr<scalar_t>();
+          scalar_t* s_values_ptr = s_values.mutable_data_ptr<scalar_t>();
```
```suggestion
          const scalar_t* t_values_ptr = t_values.const_data_ptr<scalar_t>();
          const scalar_t* s_values_ptr = s_values.const_data_ptr<scalar_t>();
```

pearu commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -653,10 +653,10 @@ static inline void add_dense_sparse_worker_non_coalesced_cpu(Tensor& r, const Sc
   auto values_dense_size = values.stride(0);
   TORCH_CHECK(values.is_contiguous());
-  scalar_t* v_ptr = values.data_ptr<scalar_t>();
+  scalar_t* v_ptr = values.mutable_data_ptr<scalar_t>();
```
```suggestion
  const scalar_t* v_ptr = values.const_data_ptr<scalar_t>();
```
plus use
```c++
const scalar_t* v_index = v_ptr + n * values_dense_size;
```
below.

pearu commented on aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp:
```diff
@@ -139,11 +139,11 @@ void block_sparse_triangular_solve_vec(
+        auto crow_indices_data_ptr = crow_indices.mutable_data_ptr<int>();
         auto col_indices = A.col_indices().to(kInt);
-        auto col_indices_data_ptr = col_indices.data_ptr<int>();
+        auto col_indices_data_ptr = col_indices.mutable_data_ptr<int>();
```
```suggestion
        const auto* col_indices_data_ptr = col_indices.const_data_ptr<int>();
```

pearu commented on aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp:
```diff
@@ -263,11 +263,11 @@ void block_sparse_triangular_solve_mat(
+        auto crow_indices_data_ptr = crow_indices.mutable_data_ptr<int>();
         auto col_indices = A.col_indices().to(kInt);
-        auto col_indices_data_ptr = col_indices.data_ptr<int>();
+        auto col_indices_data_ptr = col_indices.mutable_data_ptr<int>();
```
```suggestion
        const auto* col_indices_data_ptr = col_indices.const_data_ptr<int>();
```

pearu commented on aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp:
```diff
@@ -381,11 +381,11 @@ void block_sparse_mv(
+        auto crow_indices_data_ptr = crow_indices.mutable_data_ptr<int>();
         auto col_indices = mat.col_indices().to(kInt);
-        auto col_indices_data_ptr = col_indices.data_ptr<int>();
+        auto col_indices_data_ptr = col_indices.mutable_data_ptr<int>();
```
Use `const auto* ... const_data_ptr ...`

pearu commented on aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp:
```diff
@@ -1027,16 +1027,16 @@ void add_out_sparse_csr(
-  auto C_col_indices_ptr = C_col_indices.data_ptr<int>();
+  auto A_col_indices_ptr = A_col_indices.mutable_data_ptr<int>();
+  auto B_col_indices_ptr = B_col_indices.mutable_data_ptr<int>();
+  auto C_col_indices_ptr = C_col_indices.mutable_data_ptr<int>();
```
Use const auto* ... const_data_ptr ...

pearu commented on aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:
```diff
@@ -65,7 +65,7 @@ namespace {
     Tensor rowIndicesInt = at::empty({rowIndices.size(0)}, CUDA(kInt));
     rowIndicesInt.copy_(rowIndices);
-    sparse::cuda::Xcoo2csr(rowIndicesInt.data_ptr<int32_t>(), nnz, dim, csr.data_ptr<int32_t>());
+    sparse::cuda::Xcoo2csr(rowIndicesInt.mutable_data_ptr<int32_t>(), nnz, dim, csr.mutable_data_ptr<int32_t>());
```
```suggestion
    sparse::cuda::Xcoo2csr(rowIndicesInt.const_data_ptr<int32_t>(), nnz, dim, csr.mutable_data_ptr<int32_t>());
```

pearu commented on aten/src/ATen/native/sparse/cuda/SparseSemiStructuredOps.cu:
```diff
@@ -945,11 +945,11 @@ _to_sparse_semi_structured(const Tensor& dense) {
       if (meta_dtype == at::kShort) {
         using MetaElement = int16_t;
-        const auto meta_cpu_ptr = meta_cpu.data_ptr<MetaElement>();
+        const auto meta_cpu_ptr = meta_cpu.mutable_data_ptr<MetaElement>();
```
```suggestion
        const auto* meta_cpu_ptr = meta_cpu.const_data_ptr<MetaElement>();
```

pearu commented on aten/src/ATen/native/vulkan/ops/Utils.cpp:
```diff
@@ -370,10 +370,10 @@ api::utils::vec4 extract_texel(const Tensor& input, const ivec3& pos) {
+      convert(v_outputs_x).cpu().mutable_data_ptr<float>()[0],
+      convert(v_outputs_y).cpu().mutable_data_ptr<float>()[0],
+      convert(v_outputs_z).cpu().mutable_data_ptr<float>()[0],
+      convert(v_outputs_w).cpu().mutable_data_ptr<float>()[0],
```
These all should use const_data_ptr, afaict.

pearu reviewed (changes requested):
I have a bunch of `mutable_data_ptr->const_data_ptr` fixes and nits.

My review does not cover changes in transformers/cuda.

cyyever commented:
@pearu Thank you for your patches. However, the number of fixes is large enough to worth a dedicated PR. Therefore, could you propose a PR consisting of them so that it's easier to review?

cyyever commented:
OK. Let's try them here

amjames reviewed (changes requested):
@cyyever I'm going to recommend this gets split up. Its not that the diff is too large, but it touches to many places.

@pearu gave the sparse module a look and found some corrections, and it is reasonable to assume that someone who is more familiar with other modules will find similar in those parts of the diff.

With pearu's comments addressed I would have no problem clicking "approve" if that was the the whole diff, others will feel the same way.

Separate PRs for each module will let different folks with the necessary context review, provide additional fixes, and merge with confidence.

NikhilAPatel committed:
[Inductor] Fix benchmark_example_value losing dtype on view unwrap (#176859)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/176859
Approved by: https://github.com/mlazos
ghstack dependencies: #176543, #176544, #176545, #176546, #176547, #176548, #176549, #176845, #176847
